core/src/main/java/io/grpc/internal/CallCredentialsApplyingTransportFactory.java

```diff
@@ -126,5 +143,50 @@ public Attributes getTransportAttrs() {
         return delegate.newStream(method, headers, callOptions);
       }
     }
+
+    @Override
+    public void shutdown(Status status) {
+      checkNotNull(status, "status");
+      synchronized (this) {
+        if (pendingApplier.get() != 0) {
+          savedShutdownStatus = status;
+          return;
+        }
+      }
+      super.shutdown(status);
+    }
+
+    // TODO(zivy@): add cancel pending applier.
+    @Override
+    public void shutdownNow(Status status) {
+      checkNotNull(status, "status");
+      synchronized (this) {
+        if (pendingApplier.get() != 0) {
+          savedShutdownNowStatus = status;
+          return;
```

Comment:
This delays the majority of the channel shutdown until the applier completes, which may take a long time. We should really call `super.shutdownNow()` immediately, but then we have to manage the listener `transportTerminated()` as well.

Since this is fixing a correctness issue, doing this in this PR could be okay, but let's at least put a TODO to propagate `shutdownNow()` immediately.

Reply:
This TODO is still missing.

Reply:
Oops. I missed the "should call delegate shutdownNow asap. Maybe cancel pending applier" comment above. This was fine as it was, but this is also fine.